### PR TITLE
Change the SnowflakeId from static singleton to dependency injection singleton #1322

### PR DESCRIPTION
**Pull request by 3ldar**
### Description:
Converts `SnowflakeId` from a singleton object into a singleton service that can be altered by the users.

#### Issue(s) addressed:
- #1322

#### Changes:
- It might introduce a breaking change if someone is using the `SnowflakeId.Default()` method.

#### How to test:
- In the startup, after calling `AddCap` to register cap services a user can add a singleton service that implements `ISnowlakeId` implements to change how the cap message Ids are generated.

### Additional notes (optional):
_Provide any additional notes or context that may be relevant to the changes._

### Checklist:
- [x] I have tested my changes locally
- [ ] I have added the necessary documentation (if applicable)
- [x] I have updated the relevant tests (if applicable)
- [x] My changes follow the project's code style guidelines

### Reviewers:
- @yang-xiaodong

### CODE REVIEW

**Comment by yang-xiaodong** [comment]
The CustomHeaders option is present in almost every Transport, and I see that you added a `CustomHeadersBuilder` in AzureServiceBus, It doesn't look pretty. so I'm wondering if it's worth making such a big change for this? and you now using Redis to set environment variables doesn't looks bad, let me know what do you think?

**Comment by 3ldar** [comment]
> Hi @3ldar ,
> 
> Thanks for your PR!
> 
> The CustomHeaders option is present in almost every Transport, and I see that you added a `CustomHeadersBuilder` in AzureServiceBus, It doesn't look pretty. so I'm wondering if it's worth making such a big change for this? and you now using Redis to set environment variables doesn't looks bad, let me know what do you think?

I know it works and serves the purpose. But as a general design note, I don't like to have singleton-like anti-patterns in my code where I can have a singleton service instead. To me, it's worth making all the `CustomHeadersBuilder`s to provide an `IServiceProvider` instance because in some scenarios you will eventually need some services (Like a DbContext, or a business service) and to be able to make it work you will need to build services to get the service provider, which is not recommended and also will mess up your already registered singleton services, besides it might be even impossible when a service depends on a cap service (like `ICapPublisher`) and you need that service for the custom headers generation. So will add an alternate for all the `CustomHeaders` in the options.

File: src/DotNetCore.CAP.AzureServiceBus/CAP.AzureServiceBusOptions.cs [hunk 1]
```diff
@@ -62,13 +65,22 @@ public class AzureServiceBusOptions
...
         public Func<ServiceBusReceivedMessage, List<KeyValuePair<string, string>>>? CustomHeaders { get; set; }
 
+        /// <summary>
+        /// Use this function to write additional headers from the original ASB Message or any Custom Header, i.e. to allow compatibility with heterogeneous systems, into <see cref="CapHeader"/>       
+        /// </summary>
+        /// <remarks>
+        /// This property is a copy cat of <see cref="CustomHeaders"/>. The difference is: It also accpets a <see cref="IServiceProvider"/> to use DI services.
+        /// This is added due to converting <see cref="SnowflakeId"/> into a singleton service.
+        /// </remarks>
+        public Func<ServiceBusReceivedMessage, IServiceProvider, List<KeyValuePair<string, string>>>? CustomHeadersBuilder { get; set; }
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP.AzureServiceBus/CAP.AzureServiceBusOptions.cs:75]: I suggest Add `[Obsolete("xxxxx")]` for old `CustomHeaders` option, we will  remove old option at the feature

File: src/DotNetCore.CAP.Kafka/CAP.KafkaOptions.cs [hunk 2]
```diff
@@ -48,6 +48,11 @@ public KafkaOptions()
         /// </summary>
         public Func<ConsumeResult<string, byte[]>, List<KeyValuePair<string, string>>>? CustomHeaders { get; set; }
 
+        /// <summary>
+        /// If you need to get offset and partition and so on.., you can use this function to write additional header into <see cref="CapHeader"/>
+        /// </summary>
+        public Func<ConsumeResult<string, byte[]>,IServiceProvider, List<KeyValuePair<string, string>>>? CustomHeadersBuilder { get; set; }
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP.Kafka/CAP.KafkaOptions.cs:54]: I suggest Add `[Obsolete("xxxxx")]` for old `CustomHeaders` option, we will  remove old option at the feature

File: src/DotNetCore.CAP.RabbitMQ/CAP.RabbiMQOptions.cs [hunk 3]
```diff
@@ -84,6 +85,11 @@ public class RabbitMQOptions
         /// </summary>
         public Func<BasicDeliverEventArgs, List<KeyValuePair<string, string>>>? CustomHeaders { get; set; }
 
+        /// <summary>
+        /// If you need to get additional native delivery args, you can use this function to write into <see cref="CapHeader"/>.
+        /// </summary>
+        public Func<BasicDeliverEventArgs, IServiceProvider, List<KeyValuePair<string, string>>>? CustomHeadersBuilder { get; set; }
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP.RabbitMQ/CAP.RabbiMQOptions.cs:91]: I suggest Add `[Obsolete("xxxxx")]` for old `CustomHeaders` option, we will  remove old option at the feature

File: src/DotNetCore.CAP/Internal/SnowflakeId.Default.cs [hunk 4]
```diff
@@ -71,30 +67,15 @@ public SnowflakeId(long workerId)
...
-    public static SnowflakeId Default()
-    {
-        if (_snowflakeId != null) return _snowflakeId;
-
-        lock (SLock)
-        {
-            if (_snowflakeId != null) return _snowflakeId;
-
-            if (!long.TryParse(Environment.GetEnvironmentVariable("CAP_WORKERID"), out var workerId))
-                workerId = Util.GenerateWorkerId(MaxWorkerId);
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP/Internal/SnowflakeId.Default.cs:85]: Cannot be removed, placed into the constructor as the default behavior

File: src/DotNetCore.CAP/CAP.ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -29,7 +31,7 @@ public static CapBuilder AddCap(this IServiceCollection services, Action<CapOpti
 
         services.AddSingleton(_ => services);
         services.TryAddSingleton(new CapMarkerService("CAP"));
-
+        services.AddSingleton<ISnowflakeId>(_ => new SnowflakeId(Util.GenerateWorkerId(1023)));
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP/CAP.ServiceCollectionExtensions.cs:34]: I suggest here
```
services.AddSingleton<ISnowflakeId,SnowflakeId>();
```

File: src/DotNetCore.CAP/Internal/SnowflakeId.Default.cs (hunk 4 above)
Comment (3ldar) [reply src/DotNetCore.CAP/Internal/SnowflakeId.Default.cs:85]: Added a parameterless constructor that utilizes the above method.

**Review by yang-xiaodong — APPROVED** [review]
Thanks for the PR, since this is a Breaking Changes we will merge in the next minor release

**Pull request merged**